File: api/transactions.py
import services.transaction_service as service
from api.responses import safe


def get_transactions_by_user(user_id):
    return safe(lambda: service.get_transactions_by_user(user_id))


def get_transactions_by_account(account_id):
    return safe(lambda: service.get_transactions_by_account(account_id))


def get_transaction(transaction_id):
    return safe(lambda: service.get_transaction(transaction_id))


def create_transaction(account_id, amount, transaction_date, category_id,
                       budget_item_id=None, payee=None, notes=None,
                       recurring_rule_id=None):
    return safe(lambda: service.create_transaction(
        account_id=account_id,
        amount=amount,
        transaction_date=transaction_date,
        category_id=category_id,
        budget_item_id=budget_item_id,
        payee=payee,
        notes=notes,
        recurring_rule_id=recurring_rule_id,
    ))


def create_transfer(from_account_id, to_account_id, amount, transaction_date,
                    category_id, payee="Transfer", notes=None):
    return safe(lambda: service.create_transfer(
        from_account_id=from_account_id,
        to_account_id=to_account_id,
        amount=amount,
        transaction_date=transaction_date,
        category_id=category_id,
        payee=payee,
        notes=notes,
    ))


def update_transaction(transaction_id, account_id=None, amount=None,
                       transaction_date=None, category_id=None,
                       budget_item_id=None, payee=None, notes=None,
                       recurring_rule_id=None):
    return safe(lambda: service.update_transaction(
        transaction_id,
        account_id=account_id,
        amount=amount,
        transaction_date=transaction_date,
        category_id=category_id,
        budget_item_id=budget_item_id,
        payee=payee,
        notes=notes,
        recurring_rule_id=recurring_rule_id,
    ))


def delete_transaction(transaction_id):
    return safe(lambda: service.delete_transaction(transaction_id))


def update_transfer(transfer_id, amount=None, transaction_date=None,
                    payee=None, notes=None):
    return safe(lambda: service.update_transfer(
        transfer_id,
        amount=amount,
        transaction_date=transaction_date,
        payee=payee,
        notes=notes,
    ))


def delete_transfer(transfer_id):
    return safe(lambda: service.delete_transfer(transfer_id))

File: frontend/src/lib/api/transactions.ts
import { callApi } from './client';

export type Transaction = {
	transaction_id: number;
	account_id: number;
	category_id: number;
	budget_item_id?: number | null;
	recurring_rule_id?: number | null;
	payee?: string | null;
	notes?: string | null;
	amount_minor: number;
	transaction_date: string;
	created_at: string;
	transfer_id?: number | null;
};

export function getTransactionsByUser(userId: number) {
	return callApi<Transaction[]>('get_transactions_by_user', userId);
}

export function getTransactionsByAccount(accountId: number) {
	return callApi<Transaction[]>('get_transactions_by_account', accountId);
}

export function getTransaction(transactionId: number) {
	return callApi<Transaction | null>('get_transaction', transactionId);
}

export function createTransaction(payload: {
	account_id: number;
	amount: number | string;
	transaction_date: string;
	category_id: number;
	budget_item_id?: number | null;
	payee?: string | null;
	notes?: string | null;
	recurring_rule_id?: number | null;
}) {
	return callApi<{ transaction_id: number }>(
		'create_transaction',
		payload.account_id,
		payload.amount,
		payload.transaction_date,
		payload.category_id,
		payload.budget_item_id ?? null,
		payload.payee ?? null,
		payload.notes ?? null,
		payload.recurring_rule_id ?? null
	);
}

export function createTransfer(payload: {
	from_account_id: number;
	to_account_id: number;
	amount: number | string;
	transaction_date: string;
	category_id: number;
	payee?: string | null;
	notes?: string | null;
}) {
	return callApi<{
		transfer_id: number;
		debit_transaction_id: number;
		credit_transaction_id: number;
	}>(
		'create_transfer',
		payload.from_account_id,
		payload.to_account_id,
		payload.amount,
		payload.transaction_date,
		payload.category_id,
		payload.payee ?? 'Transfer',
		payload.notes ?? null
	);
}

export function updateTransaction(
	transactionId: number,
	payload: {
		account_id?: number | null;
		amount?: number | string | null;
		transaction_date?: string | null;
		category_id?: number | null;
		budget_item_id?: number | null;
		payee?: string | null;
		notes?: string | null;
		recurring_rule_id?: number | null;
	} = {}
) {
	return callApi<{ status: string }>(
		'update_transaction',
		transactionId,
		payload.account_id ?? null,
		payload.amount ?? null,
		payload.transaction_date ?? null,
		payload.category_id ?? null,
		payload.budget_item_id ?? null,
		payload.payee ?? null,
		payload.notes ?? null,
		payload.recurring_rule_id ?? null
	);
}

export function deleteTransaction(transactionId: number) {
	return callApi<{ status: string }>('delete_transaction', transactionId);
}

export function updateTransfer(
	transferId: number,
	payload: {
		amount?: number | string | null;
		transaction_date?: string | null;
		payee?: string | null;
		notes?: string | null;
	} = {}
) {
	return callApi<{ status: string }>(
		'update_transfer',
		transferId,
		payload.amount ?? null,
		payload.transaction_date ?? null,
		payload.payee ?? null,
		payload.notes ?? null
	);
}

export function deleteTransfer(transferId: number) {
	return callApi<{ status: string }>('delete_transfer', transferId);
}
